torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -510,3 +511,109 @@ def shard_parameter(
 
     # Now we can set the attribute appropriately.
     setattr(module, param_name, st)
+
+
+def init_sharded_tensor_from_local_result(
+    sharded_tensor,
+    local_result,
+    tensor_shard_dim,
+    result_shard_dim,
+    world_size,
+    pg,
+):
+    """
+    Given a sharded tensor and local_result from an op on top of it. We want
+    to create a new sharded tensor from the local_result so that the the next
+    op can be performed on the basis of the new sharded tensor. This can seen
+    as the last step of the first phase of the Megatron-LM style model(tensor)
+    parallelism.
+
+    Args:
+        sharded_tensor: Sharded tensor which the op was performed on.
+        local_result: A tensor which is from the op performed on the local_shard of
+            the sharded_tensor.
+        tensor_shard_dim: Dim which the tensor is sharded on.
+        result_shard_dim: Dim which the new sharded tensor will be sharded on.
+        world_size: number of ranks.
+        pg (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+
+    Return: new sharded tensor from the local_result.
+    """
+    sharded_weight_metadata = sharded_tensor.local_shards()[0].metadata
+    current_offsets = [0] * len(local_result.size())
+    current_offsets[result_shard_dim] = sharded_weight_metadata.shard_offsets[tensor_shard_dim]
+    local_shard_metadata = ShardMetadata(
+        shard_offsets=current_offsets,
+        shard_sizes=list(local_result.size()),
+        placement=sharded_weight_metadata.placement,
+    )
+    local_shards = [Shard(local_result, local_shard_metadata)]
+    global_size = list(local_result.size())
+    global_size[result_shard_dim] = sharded_tensor.size(tensor_shard_dim)
+
+    new_st = ShardedTensor._init_from_local_shards(local_shards, tuple(global_size), process_group=pg)
+
+    # Manually set sharding_spec
+    new_st._sharding_spec = copy.deepcopy(sharded_tensor._sharding_spec)
+    new_st._sharding_spec.dim = result_shard_dim
+    return new_st
+
+
+def merge_sharded_local_results(
+    sharded_local_result,
+    world_size,
+    pg=None,
+):
+    """
```

Comment:
This is ok for a prototype, but for a proper API we should have something like this:

```
class ShardedTensor(...):
  def reshard(self, 
      sharding_spec: ShardingSpec, 
      output_local_tensors: bool = False) -> Union[Tensor/ShardedTensor]:
    """
    Reshards a tensor given the sharding_spec, if output_local_tensors is True. 
    It would output the local shard on  each rank instead of ShardedTensor 
     (only works for a single local shard).
    """

def reshard_output(
      module: nn.Module, 
      sharding_spec: ShardingSpec, 
      output_local_tensors: bool = False) -> nn.Module:
  """
  Replaces a module with a new module that will reshard its output.
  """

class ModuleResharder(nn.Module):
    
   def __init__(self, original_module, sharding_spec, output_local_tensors):
       self.original_module = original_module
       self.sharding_spec = sharding_spec
       self.output_local_tensors = output_local_tensors

   def forward(self, *args, **kwargs):
     output = self.original_module(*args, **kwargs)
     return output.reshard(self.sharding_spec, self.output_local_tensors)
     
# Now code for sharded linear
ChunkShardingSpec(
            dim=sharding_dim,
            placements=[
                "rank:0/cuda:0",
                "rank:1/cuda:1",
                "rank:2/cuda:2",
                "rank:3/cuda:3",
            ],
        ),

fc = nn.Linear(10, 10)
shard_parameter(fc, "weight", spec)

fc = reshard_output(fc, resharding_spec, output_local_tensors=True)

result = fc(torch.rand(10))
```

Reply:
cc @wanchaol Some of the unified sharder components (ex: reshard_output) are relevant here. @fduwjj Is looking into a design of what this might look like for both ShardedTensor and potentially PartialTensor and maybe you can help with some of the brainstorming around the API too :)

Reply:
Synced offline, will send out a separate PR for the whole design and update the current PR on top of that.

Reply:
API defined in this PR:  https://github.com/pytorch/pytorch/pull/70079